src/web_analysis/gpt.py

```diff
@@ -0,0 +1,263 @@
+        Makes an asynchronous request to the OpenAI Chat API to generate completions for a given prompt.
+
+        Parameters:
+        - session (aiohttp.ClientSession): A session object used to make asynchronous requests.
+        - final_prompt (str): The final prompt to be sent to the OpenAI Chat API.
+
+        Returns:
+        - str: The response from the OpenAI Chat API containing the completion for the given prompt.
+        """
+        url = "https://api.openai.com/v1/chat/completions"  # endpoint for gpt-4, gpt-4-turbo-preview, gpt-3.5-turbo
+        headers = {
+            "Authorization": f"Bearer {os.environ['OPENAI_API_KEY']}",
+            "Content-Type": "application/json",
+        }
+        payload = {
+            "model": self.model,
+            "temperature": 0.2,           # lower temperature for more deterministic outputs
+            "top_p": 0.1,                 # lower top_p to decrease randomness
+            "messages": [
+                {"role": "system", "content": self.SYSTEM_PROMPT},
+                {"role": "user", "content": self.USER_PROMPT_1},
+                {"role": "assistant", "content": self.ASSISTANT_PROMPT_1},
+                {"role": "user", "content": final_prompt}
+            ]
+        }
+        async with session.post(url, json=payload, headers=headers) as response:
+            # print(f"status code: {response.status}")
+            # status code of 200 indicates a successful connection
+            if response.status == 200:
+                data = await response.json()
+                return data['choices'][0]['message']['content'].strip(" \n")
+            else:
+                return None
+
+                                                ####### batched processing #######
+
+    async def process_batch_async(self, session, batch, custom_guidelines_prompt=None):
+        """
+        Processes a single batch of prompts asynchronously.
+
+        Parameters:
+        - session (aiohttp.ClientSession): A session object used to make asynchronous requests.
+        - batch (list of str): List of prompts to be processed in the batch.
+
+        Returns:
+        - list of str: List of responses from the OpenAI Chat API, each containing the completion for the corresponding prompt in the batch.
+        """
+        if custom_guidelines_prompt != None:
+            try:
+                tasks = [self.make_openai_request_async(session, custom_guidelines_prompt.format(prompt)) for prompt in batch]
+            except Exception as e:
+                print(f"Failed to format custom guidlines. Make sure guidelines are formatted with empty brackets for where prompts will be inerted. {e}")
+        else:
+            tasks = [self.make_openai_request_async(session, self.GUIDELINES_PROMPT_TEMPLATE.format(prompt)) for prompt in batch]
+        return await asyncio.gather(*tasks)
+
+    async def process_prompts_in_batches_async(self, prompts, batch_size=10, custom_guidelines_prompt=None):
```

Comment:
I wonder if we should wrap any of this in a cache -- i.e. a file/folder that maps inputs to outputs. If the exact same input has already been asked we simply look it up rather than querying ChatGPT?

Only if we are hitting it with a lot of repeats, which could be conceivable if TOS language is replicated in many places

Reply:
@shayne-longpre Thanks for the comments! I'm working on modifying the prompts today and can add a cache to save responses soon